File: server/models/userModel.js
const mongoose = require("mongoose");

const paymentTransactionsSchema = new mongoose.Schema({
  id: { type: String, required: true },
  userId: { type: mongoose.Schema.Types.ObjectId, ref: "User", required: true },
  orderId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: "Order",
    required: true,
  },
  paymentMethod: { type: String, required: true },
  paymentDate: { type: Date, default: Date.now },
  paymentStatus: { type: String, required: true },
  paymentReference: { type: String },
});
const Role = {
  CLIENT: "client",
  ADMIN: "admin",
};
const addressSchema = new mongoose.Schema({
  location: String,
  detail: String,
  district_id: Number,
  ward_code: Number,
  isSelected: {
    type: Boolean,
    default: false,
  },
});

const userSchema = new mongoose.Schema({
  userName: { type: String },
  email: { type: String, required: true, unique: true },
  display_name: { type: String },
  password: { type: String, required: true },
  avatar: { type: String },
  address: [addressSchema],
  addressSelected: {
    location: String,
    detail: String,
    district_id: Number,
    ward_code: Number,
  },
  phoneNumber: { type: Number },
  favorites: {
    type: Array,
    default: [],
  },
  payment_transaction: [paymentTransactionsSchema],
  device_token: { type: String },
  createdAt: { type: Date, default: Date.now },
  role: {
    type: String,
    enum: Object.values(Role),
    default: Role.CLIENT,
  },
  recent_products: [{ type: mongoose.Schema.Types.ObjectId, ref: "Product" }],
});

const User = mongoose.model("User", userSchema);

module.exports = {
  User,
  paymentTransactionsSchema,
  addressSchema,
};
//add

File: admin_panel/lib/feature/user/model/user_model.dart
class User {
  final String? role;
  final String? id;
  final String? userName;
  final String? email;
  final String? password;
  final List<String>? favorites;
  final List<Address>? address;
  final List<String>? paymentTransaction;
  final DateTime? createdAt;
  final int? version;
  final String? deviceToken;
  final String? avatar;
  final String? displayName;
  final int? phoneNumber;
  final List<String>? recentProducts;
  final AddressSelected? addressSelected;

  User({
    this.role,
    this.id,
    this.userName,
    this.email,
    this.password,
    this.favorites,
    this.address,
    this.paymentTransaction,
    this.createdAt,
    this.version,
    this.deviceToken,
    this.avatar,
    this.displayName,
    this.phoneNumber,
    this.recentProducts,
    this.addressSelected,
  });

  factory User.fromJson(Map<String, dynamic> json) {
    return User(
      role: json['role'],
      id: json['_id'],
      userName: json['userName'],
      email: json['email'],
      password: json['password'],
      favorites: json['favorites'] != null ? List<String>.from(json['favorites']) : null,
      address: json['address'] != null
          ? (json['address'] as List).map((item) => Address.fromJson(item)).toList()
          : null,
      paymentTransaction: json['payment_transaction'] != null
          ? List<String>.from(json['payment_transaction'])
          : null,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      version: json['__v'],
      deviceToken: json['device_token'],
      avatar: json['avatar'],
      displayName: json['display_name'],
      phoneNumber: json['phoneNumber'],
      recentProducts: json['recent_products'] != null
          ? List<String>.from(json['recent_products'])
          : null,
      addressSelected: json['addressSelected'] != null
          ? AddressSelected.fromJson(json['addressSelected'])
          : null,
    );
  }
}

class Address {
  final String? location;
  final String? detail;
  final bool? isSelected;
  final String? id;

  Address({
    this.location,
    this.detail,
    this.isSelected,
    this.id,
  });

  factory Address.fromJson(Map<String, dynamic> json) {
    return Address(
      location: json['location'],
      detail: json['detail'],
      isSelected: json['isSelected'],
      id: json['_id'],
    );
  }
}

class AddressSelected {
  final String? location;
  final String? detail;
  final int? districtId;
  final int? wardCode;

  AddressSelected({
    this.location,
    this.detail,
    this.districtId,
    this.wardCode,
  });

  factory AddressSelected.fromJson(Map<String, dynamic> json) {
    return AddressSelected(
      location: json['location'],
      detail: json['detail'],
      districtId: json['district_id'],
      wardCode: json['ward_code'],
    );
  }
}